File: services/data-retention.js
/**
 * Data retention service -- scheduled purge of expired PHI data.
 *
 * HIPAA requires defined retention periods for protected health information.
 * This service runs once per day (integrated into the Node.js scheduler) and
 * deletes data older than the configured retention period for each table.
 *
 * Retention periods are configurable via environment variables.
 * Purges use batched deletes via CTEs to avoid long-running transactions.
 */

import { db } from '../db/client.js';
import { sql } from 'drizzle-orm';
import { createLogger } from '../lib/logger.js';

const log = createLogger('DataRetention');

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const BATCH_SIZE = 5000;

/**
 * Retention periods in days, configurable via environment variables.
 * Set a value to 0 to disable purge for that table.
 */
const RETENTION_DAYS = {
  conversation_phi:      parseInt(process.env.RETENTION_CONVERSATIONS_DAYS              || '365', 10),
  conversations:         parseInt(process.env.RETENTION_CONVERSATION_METADATA_DAYS      || '1095', 10),
  memories:              parseInt(process.env.RETENTION_MEMORIES_DAYS                   || '730', 10),
  call_analyses:         parseInt(process.env.RETENTION_CALL_ANALYSES_DAYS              || '365', 10),
  daily_call_context:    parseInt(process.env.RETENTION_DAILY_CONTEXT_DAYS              || '90',  10),
  call_metrics:          parseInt(process.env.RETENTION_CALL_METRICS_DAYS               || '180', 10),
  inactive_reminders:    parseInt(process.env.RETENTION_INACTIVE_REMINDERS_DAYS         || '365', 10),
  expired_onetime_reminders: parseInt(process.env.RETENTION_EXPIRED_ONETIME_REMINDERS_DAYS || '1', 10),
  reminder_deliveries:   parseInt(process.env.RETENTION_REMINDER_DELIVERIES_DAYS        || '90',  10),
  notifications:         parseInt(process.env.RETENTION_NOTIFICATIONS_DAYS              || '180', 10),
  caregiver_notes:       parseInt(process.env.RETENTION_CAREGIVER_NOTES_DAYS            || '365', 10),
  prospects:             parseInt(process.env.RETENTION_PROSPECTS_DAYS                  || '90',  10),
  senior_profile_review: parseInt(process.env.RETENTION_INACTIVE_SENIOR_REVIEW_DAYS     || '365', 10),
  idempotency_keys:      parseInt(process.env.RETENTION_IDEMPOTENCY_KEYS_DAYS           || '1',   10),
  call_queue:            parseInt(process.env.RETENTION_CALL_QUEUE_DAYS                 || '90',  10),
  call_attempts:         parseInt(process.env.RETENTION_CALL_ATTEMPTS_DAYS              || '180', 10),
  post_call_jobs:        parseInt(process.env.RETENTION_POST_CALL_JOBS_DAYS             || '180', 10),
  outbound_call_guards:  parseInt(process.env.RETENTION_OUTBOUND_CALL_GUARDS_DAYS       || '30',  10),
  scheduler_shadow_comparisons: parseInt(process.env.RETENTION_SCHEDULER_SHADOW_COMPARISONS_DAYS || '30', 10),
  waitlist:              parseInt(process.env.RETENTION_WAITLIST_DAYS                   || '365', 10),
  audit_logs:            parseInt(process.env.RETENTION_AUDIT_LOGS_DAYS                 || '2190', 10),
};

/**
 * Map of table name -> date column used for age comparison.
 * Only tables listed here will be purged.
 */
const TABLE_DATE_COLUMNS = {
  conversations:        'started_at',
  memories:             'created_at',
  call_analyses:        'created_at',
  daily_call_context:   'call_date',
  call_metrics:         'created_at',
  reminder_deliveries:  'created_at',
  notifications:        'sent_at',
  caregiver_notes:      'created_at',
  prospects:            'updated_at',
  call_queue:           'created_at',
  call_attempts:        'created_at',
  post_call_jobs:       'created_at',
  outbound_call_guards: 'created_at',
  scheduler_shadow_comparisons: 'created_at',
  waitlist:             'created_at',
  audit_logs:           'created_at',
};

const ALLOWED_TABLES = new Set(Object.keys(TABLE_DATE_COLUMNS));

const LEGAL_HOLD_RESOURCE_TYPES = {
  conversations: 'conversation',
  memories: 'memory',
  call_analyses: 'call_analysis',
  daily_call_context: 'daily_call_context',
  call_metrics: 'call_metric',
  reminder_deliveries: 'reminder_delivery',
  notifications: 'notification',
  caregiver_notes: 'caregiver_note',
  prospects: 'prospect',
  call_queue: 'call_queue',
  call_attempts: 'call_attempt',
  post_call_jobs: 'post_call_job',
  outbound_call_guards: 'outbound_call_guard',
  scheduler_shadow_comparisons: 'scheduler_shadow_comparison',
  waitlist: 'waitlist',
  audit_logs: 'audit_log',
};

function legalHoldExclusion(alias, table) {
  const resourceType = LEGAL_HOLD_RESOURCE_TYPES[table];
  if (!resourceType) return sql`TRUE`;
  return sql`
    NOT EXISTS (
      SELECT 1
      FROM legal_holds lh
      WHERE lh.released_at IS NULL
        AND lh.resource_type = ${resourceType}
        AND lh.resource_id = ${sql.raw(`${alias}.id::text`)}
    )
  `;
}

// ---------------------------------------------------------------------------
// Core purge logic
// ---------------------------------------------------------------------------

/**
 * Purge rows from a single table older than `days` days, in batches.
 * @param {string} table - Table name (must be in ALLOWED_TABLES)
 * @param {string} dateColumn - Column to compare against
 * @param {number} days - Retention period in days
 * @returns {Promise<number>} Total number of rows deleted
 */
async function purgeTable(table, dateColumn, days) {
  if (!ALLOWED_TABLES.has(table)) {
    throw new Error(`Table ${table} is not in ALLOWED_TABLES`);
  }

  let totalDeleted = 0;

  while (true) {
    // The table and column names come from our own hardcoded config (not user
    // input), so embedding them in the query string is safe.
    const result = await db.execute(sql`
      WITH batch AS (
        SELECT target.ctid
        FROM ${sql.raw(table)} AS target
        WHERE target.${sql.raw(dateColumn)} < NOW() - make_interval(days => ${days})
          AND ${legalHoldExclusion('target', table)}
        ORDER BY target.${sql.raw(dateColumn)}
        LIMIT ${BATCH_SIZE}
      ),
      deleted AS (
        DELETE FROM ${sql.raw(table)} AS target
        USING batch
        WHERE target.ctid = batch.ctid
        RETURNING 1
      )
      SELECT count(*)::int AS count FROM deleted
    `);

    const batchCount = result.rows?.[0]?.count ?? 0;
    totalDeleted += batchCount;

    // If we deleted fewer than the batch size, we're done.
    if (batchCount < BATCH_SIZE) break;

    // Yield between batches to avoid blocking the event loop.
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return totalDeleted;
}

/**
 * Null PHI-bearing conversation fields older than the transcript retention
 * period while preserving non-PHI metadata for longer analytics/compliance.
 */
async function redactConversationPhi(days) {
  let totalRedacted = 0;

  while (true) {
    const result = await db.execute(sql`
      WITH batch AS (
        SELECT target.ctid
        FROM conversations AS target
        WHERE target.started_at < NOW() - make_interval(days => ${days})
          AND (
            target.summary IS NOT NULL
            OR target.summary_encrypted IS NOT NULL
            OR target.transcript IS NOT NULL
            OR target.transcript_encrypted IS NOT NULL
            OR target.transcript_text_encrypted IS NOT NULL
            OR target.concerns IS NOT NULL
          )
          AND ${legalHoldExclusion('target', 'conversations')}
        ORDER BY target.started_at
        LIMIT ${BATCH_SIZE}
      ),
      redacted AS (
        UPDATE conversations AS target
        SET summary = NULL,
            summary_encrypted = NULL,
            transcript = NULL,
            transcript_encrypted = NULL,
            transcript_text_encrypted = NULL,
            concerns = NULL
        FROM batch
        WHERE target.ctid = batch.ctid
        RETURNING 1
      )
      SELECT count(*)::int AS count FROM redacted
    `);

    const batchCount = result.rows?.[0]?.count ?? 0;
    totalRedacted += batchCount;
    if (batchCount < BATCH_SIZE) break;
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return totalRedacted;
}

/**
 * Purge inactive reminder definitions after their deactivation retention window.
 */
async function purgeInactiveReminders(days) {
  let totalDeleted = 0;

  while (true) {
    const result = await db.execute(sql`
      WITH batch AS (
        SELECT r.id
        FROM reminders r
        WHERE r.is_active = false
          AND COALESCE(r.deactivated_at, r.last_delivered_at, r.created_at) < NOW() - make_interval(days => ${days})
          AND NOT EXISTS (
            SELECT 1
            FROM legal_holds lh
            WHERE lh.released_at IS NULL
              AND lh.resource_type = 'reminder'
              AND lh.resource_id = r.id::text
          )
          AND NOT EXISTS (
            SELECT 1
            FROM reminder_deliveries rd
            JOIN legal_holds lh
              ON lh.released_at IS NULL
             AND lh.resource_type = 'reminder_delivery'
             AND lh.resource_id = rd.id::text
            WHERE rd.reminder_id = r.id
          )
        ORDER BY COALESCE(r.deactivated_at, r.last_delivered_at, r.created_at)
        LIMIT ${BATCH_SIZE}
      ),
      deleted_deliveries AS (
        DELETE FROM reminder_deliveries rd
        USING batch
        WHERE rd.reminder_id = batch.id
        RETURNING 1
      ),
      deleted_reminders AS (
        DELETE FROM reminders r
        USING batch
        WHERE r.id = batch.id
        RETURNING 1
      )
      SELECT count(*)::int AS count FROM deleted_reminders
    `);

    const batchCount = result.rows?.[0]?.count ?? 0;
    totalDeleted += batchCount;
    if (batchCount < BATCH_SIZE) break;
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return totalDeleted;
}

/**
 * Delete one-time reminders whose scheduled event passed more than `days` days
 * ago. Recurring reminders are never touched. Deletes associated
 * reminder_deliveries first (no FK cascade in schema). Respects legal_holds.
 *
 * Designed to expire one-time reminders the day after their event so the
 * mobile UI doesn't keep showing past events.
 */
async function purgeExpiredOnetimeReminders(days) {
  let totalDeleted = 0;

  while (true) {
    const result = await db.execute(sql`
      WITH batch AS (
        SELECT r.id
        FROM reminders r
        WHERE r.is_recurring = false
          AND r.scheduled_time IS NOT NULL
          AND r.scheduled_time < NOW() - make_interval(days => ${days})
          AND NOT EXISTS (
            SELECT 1
            FROM legal_holds lh
            WHERE lh.released_at IS NULL
              AND lh.resource_type = 'reminder'
              AND lh.resource_id = r.id::text
          )
          AND NOT EXISTS (
            SELECT 1
            FROM reminder_deliveries rd
            JOIN legal_holds lh
              ON lh.released_at IS NULL
             AND lh.resource_type = 'reminder_delivery'
             AND lh.resource_id = rd.id::text
            WHERE rd.reminder_id = r.id
          )
        ORDER BY r.scheduled_time
        LIMIT ${BATCH_SIZE}
      ),
      deleted_deliveries AS (
        DELETE FROM reminder_deliveries rd
        USING batch
        WHERE rd.reminder_id = batch.id
        RETURNING 1
      ),
      deleted_reminders AS (
        DELETE FROM reminders r
        USING batch
        WHERE r.id = batch.id
        RETURNING 1
      )
      SELECT count(*)::int AS count FROM deleted_reminders
    `);

    const batchCount = result.rows?.[0]?.count ?? 0;
    totalDeleted += batchCount;
    if (batchCount < BATCH_SIZE) break;
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return totalDeleted;
}

/**
 * Count inactive senior profiles ready for manual review. This intentionally
 * does not delete profile rows because policy requires review before purge.
 */
async function countSeniorProfileReviewCandidates(days) {
  const result = await db.execute(sql`
    SELECT COUNT(*)::int AS count
    FROM seniors s
    WHERE s.is_active = false
      AND COALESCE(
        (SELECT MAX(c.started_at) FROM conversations c WHERE c.senior_id = s.id),
        s.updated_at,
        s.created_at
      ) < NOW() - make_interval(days => ${days})
      AND NOT EXISTS (
        SELECT 1
        FROM legal_holds lh
        WHERE lh.released_at IS NULL
          AND lh.resource_type = 'senior'
          AND lh.resource_id = s.id::text
      )
  `);
  return result.rows?.[0]?.count ?? 0;
}

/**
 * Delete idempotency replay cache entries after their explicit expiration.
 * Cached responses are encrypted, but they can still contain PHI after
 * authorized decrypt-at-boundary replay, so they must not outlive their TTL.
 */
async function purgeExpiredIdempotencyKeys() {
  let totalDeleted = 0;

  while (true) {
    const result = await db.execute(sql`
      WITH batch AS (
        SELECT ctid
        FROM idempotency_keys
        WHERE expires_at < NOW()
        ORDER BY expires_at
        LIMIT ${BATCH_SIZE}
      ),
      deleted AS (
        DELETE FROM idempotency_keys AS target
        USING batch
        WHERE target.ctid = batch.ctid
        RETURNING 1
      )
      SELECT count(*)::int AS count FROM deleted
    `);

    const batchCount = result.rows?.[0]?.count ?? 0;
    totalDeleted += batchCount;
    if (batchCount < BATCH_SIZE) break;
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return totalDeleted;
}

/**
 * Delete rows older than their retention period from all PHI tables.
 * @returns {Promise<Object>} Map of table name -> number of rows deleted
 */
export async function purgeExpiredData() {
  const results = {};

  for (const [table, days] of Object.entries(RETENTION_DAYS)) {
    if (days <= 0) continue; // 0 = disabled

    try {
      if (table === 'conversation_phi') {
        results[table] = await redactConversationPhi(days);
        continue;
      }

      if (table === 'inactive_reminders') {
        results[table] = await purgeInactiveReminders(days);
        continue;
      }

      if (table === 'expired_onetime_reminders') {
        results[table] = await purgeExpiredOnetimeReminders(days);
        continue;
      }

      if (table === 'senior_profile_review') {
        results.senior_profile_review_candidates = await countSeniorProfileReviewCandidates(days);
        continue;
      }

      if (table === 'idempotency_keys') {
        results[table] = await purgeExpiredIdempotencyKeys();
        continue;
      }

      const dateColumn = TABLE_DATE_COLUMNS[table];
      if (!dateColumn) continue;

      results[table] = await purgeTable(table, dateColumn, days);
    } catch (err) {
      // Log the error but continue with remaining tables so a single
      // missing table (e.g. audit_logs not yet created) doesn't block
      // the whole purge cycle.
      log.warn(`Failed to purge ${table}`, { error: err.message });
      results[table] = -1;
    }
  }

  return results;
}

// ---------------------------------------------------------------------------
// Scheduler integration
// ---------------------------------------------------------------------------

let lastPurgeDate = null;

/**
 * Run the daily purge if it hasn't been run today.
 * Designed to be called from the scheduler's polling loop (every 60s).
 * The date check ensures it only actually runs once per calendar day.
 */
export async function runDailyPurgeIfNeeded() {
  const today = new Date().toISOString().slice(0, 10);
  if (lastPurgeDate === today) return; // Already ran today

  log.info('Starting daily data retention purge');
  lastPurgeDate = today;

  try {
    const results = await purgeExpiredData();
    const total = Object.values(results).reduce((sum, v) => sum + Math.max(v, 0), 0);

    if (total > 0) {
      log.info('Data retention purge complete', { results, total });
    } else {
      log.info('Data retention purge: nothing to delete');
    }
  } catch (err) {
    log.error('Data retention purge failed', { error: err.message });
    // Reset so we retry next cycle
    lastPurgeDate = null;
  }
}

export const dataRetentionService = {
  purgeExpiredData,
  runDailyPurgeIfNeeded,
  RETENTION_DAYS,
  purgeInactiveReminders,
  purgeExpiredOnetimeReminders,
  countSeniorProfileReviewCandidates,
};

File: pipecat/services/data_retention.py
"""Data retention service -- scheduled purge of expired PHI data.

HIPAA requires defined retention periods for protected health information.
This service runs daily as a background asyncio loop and deletes data older
than the configured retention period for each table.

Retention periods are configurable via environment variables (see config.py).
Purges use batched deletes via CTEs to avoid long-running transactions
and excessive lock contention on production.
"""

from __future__ import annotations

import asyncio

from loguru import logger

from config import settings
from db.client import query_one

# ---------------------------------------------------------------------------
# Table configuration
# ---------------------------------------------------------------------------
# Each entry maps a table name to the date column used for age comparison.
# Only tables listed here (and in ALLOWED_TABLES) will be purged.

TABLE_DATE_COLUMNS: dict[str, str] = {
    "conversations": "started_at",
    "memories": "created_at",
    "prospect_memories": "created_at",
    "call_analyses": "created_at",
    "daily_call_context": "call_date",
    "call_metrics": "created_at",
    "reminder_deliveries": "created_at",
    "notifications": "sent_at",
    "caregiver_notes": "created_at",
    "prospects": "updated_at",
    "call_queue": "created_at",
    "call_attempts": "created_at",
    "post_call_jobs": "created_at",
    "outbound_call_guards": "created_at",
    "scheduler_shadow_comparisons": "created_at",
    "canary_cohort_membership": "removed_at",
    "waitlist": "created_at",
    "audit_logs": "created_at",
}

ALLOWED_TABLES = frozenset(TABLE_DATE_COLUMNS.keys())

BATCH_SIZE = 5000

LEGAL_HOLD_RESOURCE_TYPES: dict[str, str] = {
    "conversations": "conversation",
    "memories": "memory",
    "prospect_memories": "memory",
    "call_analyses": "call_analysis",
    "daily_call_context": "daily_call_context",
    "call_metrics": "call_metric",
    "reminder_deliveries": "reminder_delivery",
    "notifications": "notification",
    "caregiver_notes": "caregiver_note",
    "prospects": "prospect",
    "call_queue": "call_queue",
    "call_attempts": "call_attempt",
    "post_call_jobs": "post_call_job",
    "outbound_call_guards": "outbound_call_guard",
    "scheduler_shadow_comparisons": "scheduler_shadow_comparison",
    "canary_cohort_membership": "canary_cohort_membership",
    "waitlist": "waitlist",
    "audit_logs": "audit_log",
}


def _legal_hold_exclusion(alias: str, table: str) -> str:
    resource_type = LEGAL_HOLD_RESOURCE_TYPES.get(table)
    if not resource_type:
        return "TRUE"
    return (
        "NOT EXISTS ("
        " SELECT 1 FROM legal_holds lh"
        " WHERE lh.released_at IS NULL"
        f" AND lh.resource_type = '{resource_type}'"
        f" AND lh.resource_id = {alias}.id::text"
        ")"
    )


def _dependency_exclusion(alias: str, table: str) -> str:
    if table != "call_queue":
        return "TRUE"
    return (
        "NOT EXISTS (SELECT 1 FROM call_attempts ca WHERE ca.queue_id = target.id)"
        " AND NOT EXISTS (SELECT 1 FROM outbound_call_guards ocg WHERE ocg.queue_id = target.id)"
        " AND NOT EXISTS (SELECT 1 FROM scheduler_shadow_comparisons ssc WHERE ssc.queue_id = target.id)"
    ).replace("target.", f"{alias}.")


# ---------------------------------------------------------------------------
# Core purge logic
# ---------------------------------------------------------------------------

async def _purge_table(table: str, date_column: str, retention_days: int) -> int:
    """Delete rows from *table* older than *retention_days* in batches.

    Returns the total number of rows deleted across all batches.
    """
    assert table in ALLOWED_TABLES, f"Table {table!r} is not in ALLOWED_TABLES"

    total_deleted = 0

    while True:
        # Use a CTE so we can count the deleted rows in one round-trip.
        # The table and column names come from our own hardcoded config, not
        # user input, so the f-string is safe here.
        result = await query_one(
            f"WITH batch AS ("
            f"  SELECT target.ctid"
            f"  FROM {table} AS target"
            f"  WHERE target.{date_column} < NOW() - make_interval(days => $1)"
            f"    AND {_legal_hold_exclusion('target', table)}"
            f"    AND {_dependency_exclusion('target', table)}"
            f"  ORDER BY target.{date_column}"
            f"  LIMIT {BATCH_SIZE}"
            f"), deleted AS ("
            f"  DELETE FROM {table} AS target"
            f"  USING batch"
            f"  WHERE target.ctid = batch.ctid"
            f"  RETURNING 1"
            f") SELECT count(*) AS count FROM deleted",
            retention_days,
        )
        batch_count = result["count"] if result else 0
        total_deleted += batch_count

        # If we deleted fewer than the batch size, we're done.
        if batch_count < BATCH_SIZE:
            break

        # Yield to the event loop between batches so we don't starve callers.
        await asyncio.sleep(0.1)

    return total_deleted


async def _redact_conversation_phi(retention_days: int) -> int:
    """Null old conversation transcripts/summaries while retaining metadata."""
    total_redacted = 0

    while True:
        result = await query_one(
            f"WITH batch AS ("
            f"  SELECT target.ctid"
            f"  FROM conversations AS target"
            f"  WHERE target.started_at < NOW() - make_interval(days => $1)"
            f"    AND ("
            f"      target.summary IS NOT NULL"
            f"      OR target.summary_encrypted IS NOT NULL"
            f"      OR target.transcript IS NOT NULL"
            f"      OR target.transcript_encrypted IS NOT NULL"
            f"      OR target.transcript_text_encrypted IS NOT NULL"
            f"      OR target.concerns IS NOT NULL"
            f"    )"
            f"    AND {_legal_hold_exclusion('target', 'conversations')}"
            f"  ORDER BY target.started_at"
            f"  LIMIT {BATCH_SIZE}"
            f"), redacted AS ("
            f"  UPDATE conversations AS target"
            f"  SET summary = NULL,"
            f"      summary_encrypted = NULL,"
            f"      transcript = NULL,"
            f"      transcript_encrypted = NULL,"
            f"      transcript_text_encrypted = NULL,"
            f"      concerns = NULL"
            f"  FROM batch"
            f"  WHERE target.ctid = batch.ctid"
            f"  RETURNING 1"
            f") SELECT count(*) AS count FROM redacted",
            retention_days,
        )
        batch_count = result["count"] if result else 0
        total_redacted += batch_count
        if batch_count < BATCH_SIZE:
            break
        await asyncio.sleep(0.1)

    return total_redacted


async def _purge_inactive_reminders(retention_days: int) -> int:
    """Delete inactive reminder definitions after their retention window."""
    total_deleted = 0

    while True:
        result = await query_one(
            f"WITH batch AS ("
            f"  SELECT r.id"
            f"  FROM reminders r"
            f"  WHERE r.is_active = false"
            f"    AND COALESCE(r.deactivated_at, r.last_delivered_at, r.created_at) < NOW() - make_interval(days => $1)"
            f"    AND NOT EXISTS ("
            f"      SELECT 1 FROM legal_holds lh"
            f"      WHERE lh.released_at IS NULL"
            f"        AND lh.resource_type = 'reminder'"
            f"        AND lh.resource_id = r.id::text"
            f"    )"
            f"    AND NOT EXISTS ("
            f"      SELECT 1"
            f"      FROM reminder_deliveries rd"
            f"      JOIN legal_holds lh"
            f"        ON lh.released_at IS NULL"
            f"       AND lh.resource_type = 'reminder_delivery'"
            f"       AND lh.resource_id = rd.id::text"
            f"      WHERE rd.reminder_id = r.id"
            f"    )"
            f"  ORDER BY COALESCE(r.deactivated_at, r.last_delivered_at, r.created_at)"
            f"  LIMIT {BATCH_SIZE}"
            f"), deleted_deliveries AS ("
            f"  DELETE FROM reminder_deliveries rd"
            f"  USING batch"
            f"  WHERE rd.reminder_id = batch.id"
            f"  RETURNING 1"
            f"), deleted_reminders AS ("
            f"  DELETE FROM reminders r"
            f"  USING batch"
            f"  WHERE r.id = batch.id"
            f"  RETURNING 1"
            f") SELECT count(*) AS count FROM deleted_reminders",
            retention_days,
        )
        batch_count = result["count"] if result else 0
        total_deleted += batch_count
        if batch_count < BATCH_SIZE:
            break
        await asyncio.sleep(0.1)

    return total_deleted


async def _count_senior_profile_review_candidates(retention_days: int) -> int:
    """Count inactive senior profiles ready for manual review, without deleting."""
    result = await query_one(
        "SELECT COUNT(*) AS count"
        " FROM seniors s"
        " WHERE s.is_active = false"
        "   AND COALESCE("
        "     (SELECT MAX(c.started_at) FROM conversations c WHERE c.senior_id = s.id),"
        "     s.updated_at,"
        "     s.created_at"
        "   ) < NOW() - make_interval(days => $1)"
        "   AND NOT EXISTS ("
        "     SELECT 1 FROM legal_holds lh"
        "     WHERE lh.released_at IS NULL"
        "       AND lh.resource_type = 'senior'"
        "       AND lh.resource_id = s.id::text"
        "   )",
        retention_days,
    )
    return result["count"] if result else 0


async def purge_expired_data() -> dict[str, int]:
    """Delete rows older than their retention period from all PHI tables.

    Returns a dict mapping table name -> number of rows deleted.
    Skips tables whose retention period is set to 0 (disabled).
    """
    retention_days = {
        "conversation_phi": settings.retention_conversations_days,
        "conversations": settings.retention_conversation_metadata_days,
        "memories": settings.retention_memories_days,
        "prospect_memories": settings.retention_memories_days,
        "call_analyses": settings.retention_call_analyses_days,
        "daily_call_context": settings.retention_daily_context_days,
        "call_metrics": settings.retention_call_metrics_days,
        "inactive_reminders": settings.retention_inactive_reminders_days,
        "reminder_deliveries": settings.retention_reminder_deliveries_days,
        "notifications": settings.retention_notifications_days,
        "caregiver_notes": settings.retention_caregiver_notes_days,
        "prospects": settings.retention_prospects_days,
        "senior_profile_review": settings.retention_inactive_senior_review_days,
        "call_attempts": settings.retention_call_attempts_days,
        "post_call_jobs": settings.retention_post_call_jobs_days,
        "outbound_call_guards": settings.retention_outbound_call_guards_days,
        "scheduler_shadow_comparisons": settings.retention_scheduler_shadow_comparisons_days,
        "call_queue": settings.retention_call_queue_days,
        "canary_cohort_membership": settings.retention_canary_cohort_membership_days,
        "waitlist": settings.retention_waitlist_days,
        "audit_logs": settings.retention_audit_logs_days,
    }

    results: dict[str, int] = {}

    for table, days in retention_days.items():
        if days <= 0:
            # 0 means retention is disabled for this table -- skip.
            continue

        try:
            if table == "conversation_phi":
                results[table] = await _redact_conversation_phi(days)
                continue

            if table == "inactive_reminders":
                results[table] = await _purge_inactive_reminders(days)
                continue

            if table == "senior_profile_review":
                results["senior_profile_review_candidates"] = await _count_senior_profile_review_candidates(days)
                continue

            date_col = TABLE_DATE_COLUMNS.get(table)
            if not date_col:
                continue

            deleted = await _purge_table(table, date_col, days)
            results[table] = deleted
        except Exception as exc:
            # Log the error but continue with the remaining tables so a
            # single missing table (e.g. audit_logs not yet created) doesn't
            # block the whole purge cycle.
            logger.warning(
                "Data retention: failed to purge {table}: {err}",
                table=table,
                err=str(exc),
            )
            results[table] = -1

    return results


# ---------------------------------------------------------------------------
# Background loop
# ---------------------------------------------------------------------------

PURGE_INTERVAL_SECONDS = 24 * 60 * 60  # 24 hours


async def start_retention_loop() -> None:
    """Run the data retention purge once every 24 hours.

    Call once at startup via ``asyncio.create_task(start_retention_loop())``.
    The first purge runs 60 seconds after startup to let the DB pool warm up.
    """
    # Wait briefly so the server can finish starting up.
    await asyncio.sleep(60)

    while True:
        try:
            results = await purge_expired_data()
            total = sum(v for v in results.values() if v > 0)
            if total > 0:
                logger.info(
                    "Data retention purge complete: {results} (total={total})",
                    results=results,
                    total=total,
                )
            else:
                logger.debug("Data retention purge: nothing to delete")
        except Exception as exc:
            logger.error("Data retention purge failed: {err}", err=str(exc))

        await asyncio.sleep(PURGE_INTERVAL_SECONDS)
